### PR TITLE
ci: add golangci-lint, yamlllint and codespell

### PR DESCRIPTION
**Pull request by lianghao208**
Signed-off-by: lianghao208 <roylizard3@gmail.com>
issue: #73

### CODE REVIEW

File: .github/workflows/build_and_test.yaml
Comment (youngnick) [inline .github/workflows/build_and_test.yaml:21]: We'll need to keep an eye on this, as golangci-lint doesn't yet support Go generics stuff, so any generics we add may cause this check to fail.

Comment (youngnick) [reply .github/workflows/build_and_test.yaml:21]: I think there may be a more recent version available as well - we should go with the latest version we can find.

File: hack/golangci-lint
Comment (youngnick) [inline hack/golangci-lint:14]: We may need to bump this version as well.

**Review by youngnick — DISMISSED** [review]
This is looking great, nice work @lianghao208!

File: .github/workflows/build_and_test.yaml
Comment (lianghao208) [reply .github/workflows/build_and_test.yaml:21]: @youngnick 
Agree, referring to the [Contour golangci-lint](https://github.com/projectcontour/contour/blob/main/.github/workflows/prbuild.yaml#L20), it does not set up golang environment either. And yes, we'll need to keep an eye on this.

Thanks for your reminder, the latest version is [1.46.2](https://github.com/golangci/golangci-lint/releases/tag/v1.46.2) for now. Will bump the version.
xref: [golangci-lint/releases](https://github.com/golangci/golangci-lint/releases)

File: hack/codespell.sh
Comment (arkodg) [inline hack/codespell.sh:2]: curious why `hack/codespell.sh` seems to have a `.sh` extension but `hack/golangci-lint` doesn't

File: Makefile
Comment (arkodg) [inline Makefile:21]: can you please raise a GH issue for this

File: Makefile
Comment (arkodg) [inline Makefile:25]: is `codespell` better than [misspell](https://github.com/golangci/misspell) thats part of `golangci-lint` ?
also if we do end up using `codespell`, the target needs to be added to `lint` :)

Comment (arkodg) [reply Makefile:25]: cc @skriss

File: hack/golangci-lint
Comment (arkodg) [inline hack/golangci-lint:3]: prefer if the project installs tooling on behalf of the user as shown [here](https://github.com/envoyproxy/gateway/pull/84)
rather than using the `hack` approach

Comment (arkodg) [inline hack/golangci-lint:3]: prefer if the project installs tooling on behalf of the user as shown [here](https://github.com/envoyproxy/gateway/pull/84)
rather than using the `hack` approach

File: Makefile
Comment (lianghao208) [reply Makefile:25]: Since projectcontour/contour is using codespell as well, maybe we should refer to contour's approaching? 

Already added codespell to lint, thanks for pointing that out.

File: hack/golangci-lint
Comment (lianghao208) [reply hack/golangci-lint]: If we use `installs tooling` in Makefile instead of using `hack` approach, I think we need to wait until #84 has been merged.

On second thought, to run golangci-lint and yamllint, we simply just use `docker run` once, I don't think we need any installation steps to set up some tools first( not `go install` related). Maybe we should keep the `hack` approach in these situations, WDYT?

File: Makefile
Comment (youngnick) [reply Makefile:25]: Codespell will check comments as well.

**Comment by danehans** [comment]
@lianghao208 thanks again for the PR. When you have a moment, can you update the PR base on the feedback?

**Comment by lianghao208** [comment]
> @lianghao208 thanks again for the PR. When you have a moment, can you update the PR base on the feedback?

@danehans I've already update the PR 2 days ago, and there seems to be some details haven't been confirmed yet. 

ping @arkodg

**Comment by arkodg** [comment]
@lianghao208 raised https://github.com/envoyproxy/gateway/issues/90 to make a call on how we install tooling
for the project, the decision made in the issue should help drive the implementation of this PR

**Comment by lianghao208** [comment]
In https://github.com/envoyproxy/gateway/issues/90#issuecomment-1147745247, we reached a consensus. Let's keep the `hack` approach and make this PR merged.
@arkodg @youngnick

File: .golangci.yml
Comment (arkodg) [inline .golangci.yml:17]: looks like misspell is already enabled, why is this comment needed

File: hack/yamllint.sh
Comment (arkodg) [inline hack/yamllint.sh:12]: can you pin to a version/tag ?

File: hack/codespell.sh
Comment (arkodg) [inline hack/codespell.sh:12]: we should be using official/library images as much as we can
looking at https://hub.docker.com/r/lianghao208/codespell, looks like it was just created for this PR
without having access to the Dockerfile that created this image and this image being used by the larger community in general, im less inclined to allow this to be checked in.
We should either skip `codespell`, or go with [one](https://github.com/envoyproxy/gateway/issues/90#issue-1259340650) to install tooling
cc @danehans @youngnick

File: .github/workflows/build_and_test.yaml
Comment (lianghao208) [reply .github/workflows/build_and_test.yaml]: done.

File: .golangci.yml
Comment (lianghao208) [reply .golangci.yml]: done, already removed the comment.

File: hack/yamllint.sh
Comment (lianghao208) [reply hack/yamllint.sh]: done.

File: hack/codespell.sh
Comment (lianghao208) [reply hack/codespell.sh:12]: Can we instruct developers to install `codespell` directly from `https://github.com/codespell-project/codespell#installation` if they haven't installed yet, instead of running a docker container for them? 

In that case, we can simply remove the `docker run` step from hack shell.

xref: [contour/codespell](https://github.com/projectcontour/contour/blob/main/hack/codespell.sh)

Comment (arkodg) [reply hack/codespell.sh:12]: @danehans above steps of asking users to manually install `codespell` is not a great dev experience. This is why [one](https://github.com/envoyproxy/gateway/issues/90#issue-1259340650) has been a very pattern adopted by many projects.

Comment (arkodg) [reply hack/codespell.sh:12]: yah based on the learnings from this PR, I prefer moving to 1. Im happy to start the work for it

**Comment by danehans** [comment]
xref https://github.com/envoyproxy/gateway/issues/63

File: hack/codespell.sh
Comment (danehans) [reply hack/codespell.sh:12]: @arkodg I'm fine with pivoting to #1. Please update #90 and cc the other maintainers to ensure everyone's on the same page.

**Comment by lianghao208** [comment]
Since #102 is still in progress, can we separate this into two PRs? 
This PR is for git workflow only, once #102 has been merged, I will work on another for the dev side(Makefile).
@arkodg @danehans

File: .github/workflows/build_and_test.yaml
Comment (arkodg) [inline .github/workflows/build_and_test.yaml:37]: is there any way to reuse .codespell.skip, rather than define it twice ?

**Comment by arkodg** [comment]
@lianghao208 https://github.com/envoyproxy/gateway/pull/102 is now in, which should unblock this PR
recommend moving any tools files such as `.codespell` into `tools/<toolname>/` , TIA

**Comment by lianghao208** [comment]
> @lianghao208 #102 is now in, which should unblock this PR recommend moving any tools files such as `.codespell` into `tools/<toolname>/` , TIA

Thanks, BTW, is the image `envoyproxy/gateway-dev-builder:latest` available now? This image seems not in docker hub yet, I can only find [envoyproxy/gateway-dev](https://hub.docker.com/r/envoyproxy/gateway-dev/tags) in docker hub.

xref: [#102 ](https://github.com/envoyproxy/gateway/pull/102/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R9)

**Comment by lianghao208** [comment]
> @lianghao208 #102 is now in, which should unblock this PR recommend moving any tools files such as `.codespell` into `tools/<toolname>/` , TIA

@arkodg Thanks, already updated the PR, PTAL.

File: .github/workflows/build_and_test.yaml
Comment (lianghao208) [reply .github/workflows/build_and_test.yaml:37]: I'am afraid there is no way reusing the .codespell.skip file, shell is not supported in `with` syntax. So we can't reuse the file like `$(shell cat .codespell.skip | tr \\n ',') `

xref: [projectcontour/contour/blob/main/.github/workflows/prbuild.yaml#L38](https://github.com/projectcontour/contour/blob/main/.github/workflows/prbuild.yaml#L38)

File: tools/codespell/Dockerfile.builder
Comment (arkodg) [inline tools/codespell/Dockerfile.builder:2]: you will need to install `codespell` and `yamllint` in tools/Dockerfile.builder

File: tools/yamllint/Dockerfile.builder
Comment (arkodg) [inline tools/yamllint/Dockerfile.builder:2]: same as above

File: Makefile.targets.mk
Comment (arkodg) [inline Makefile.targets.mk:44]: these targets should call the binary directly and not any intermediate wrapper script

File: DEVELOPER.md
Comment (arkodg) [inline DEVELOPER.md:21]: please also fill out the linters section with the exact versions and installation guides, thanks

File: .yamllint
Comment (arkodg) [inline .yamllint:1]: please move these files into `tools/yamllint/`

File: .golangci.yml
Comment (arkodg) [inline .golangci.yml:1]: please move this file under `tools/golangci-lint`

File: .codespell.ignorewords
Comment (arkodg) [inline .codespell.ignorewords:1]: can you move this file into `tools/codespell`

File: .github/workflows/build_and_test.yaml
Comment (arkodg) [reply .github/workflows/build_and_test.yaml:37]: cool maybe a comment here explaining it would be great

File: tools/codespell/Dockerfile.builder
Comment (lianghao208) [reply tools/codespell/Dockerfile.builder:2]: @arkodg 
Sorry for my miscomprehension, I thought the linters need to run as docker-in-docker(cause the tools/Dockerfile.builder only providers golang base image and has docker installed)

If we install codespell and yamllint in tools/Dockerfile.builder, does it mean we need a ubuntu base image instead of golang? (python is required for yamllint and codespell)
> FROM golang:1.18.2 as builder > FROM ubuntu:focal as builder

**Comment by lianghao208** [comment]
@arkodg 
Already updated, PTAL, thanks.

ps: I think each linter runs on separate containers saves much more time than running on single tools/Dockerfile.builder to set up environment step by step. But since istio and envoy are using this method, I guess this is a widely accepted approach.

File: tools/codespell/Dockerfile.builder
Comment (lianghao208) [reply tools/codespell/Dockerfile.builder]: Refer to Istio: https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile

I made some changes.

**Comment by danehans** [comment]
@lizan thoughts regarding https://github.com/envoyproxy/gateway/pull/79#issuecomment-1155109731?

File: DEVELOPER.md
Comment (arkodg) [reply DEVELOPER.md:21]: can you follow the format used for `make`, `docker` `go` etc, by deleting the `linters` header and adding a separate one
for `golangci-lint` , `yamllint` and `codespell` so each tools has an installation guide and version bullet point

File: Makefile
Comment (arkodg) [inline Makefile:21]: please revert this, the tools/builder container is not supposed to be based on any host context

Comment (lianghao208) [reply Makefile:21]: I was wondering if we can send tools/install-python-3.8.sh from host to Dockerfile.builder like what istio did: https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile#L845

File: tools/Dockerfile.builder
Comment (arkodg) [inline tools/Dockerfile.builder:8]: imho all this seems excessive at this point
the go image is based on Debian and already contains most tools
https://github.com/docker-library/golang/blob/master/1.18/buster/Dockerfile

the install steps for the linters seem to be a one line `RUN` step which is good

the python install can be a single RUN step,  `apt install python3-pip` is enough to install `pip` and its dependencies

Comment (lianghao208) [reply tools/Dockerfile.builder]: That's a good point, I've changed the Dockerfile.builder, PTAL.

**Review by arkodg — APPROVED** [review]
LGTM, thanks for working through the review comments !

File: Makefile
Comment (lianghao208) [reply Makefile:16]: I think removing `-it`, `-i` or `-t` can fix CI, what I did in this PR is to remove `-i`, and it works well.

**Pull request merged**
